### ISSUE TITLE
Document or improve inputs to module hooks for register_backward_hook

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#7040]
The [pytorch documentation for `nn.Module.register_backward_hook`](http://pytorch.org/docs/master/nn.html#torch.nn.Module.register_backward_hook) says:

> The grad_input and grad_output may be tuples if the module has multiple inputs or outputs. The hook should not modify its arguments, but it can optionally return a new gradient with respect to input that will be used in place of grad_input in subsequent computation.

This is not quite accurate:
- `grad_output` is indeed the gradient of the loss w.r.t. the layer output. So if you have a layer l and do, say, `y = l(x) ; loss = y.sum(); loss.backward()`, you get the gradient of `loss` w.r.t. `y`. So far so good.
-  What is unexpected to most users is that `grad_input` are the inputs to the last operation in the layer. For linear layers, this is fairly complete, as the last op is `torch.addmm` multiplying the input with the weight and adding the bias. For other layers (e.g. do a Sequential, it’ll be the last op of the last layer, the inputs not even remotely related to the sequential layer’s inputs). You can see what will be used by looking at `y.grad_fn`.

This may be confusing to users, e.g. https://discuss.pytorch.org/t/exact-meaning-of-grad-input-and-grad-output/14186

So improving the docs is one thing, but should we make the hook actually live up to the description?
I see two strategies:
- The straightforward way: collect the grad_ins with variable hooks and call the module hook when we have all of them. We loose the ability to return a different gradient.
- The somewhat convoluted way: wrap the module forward in a autograd function - similar to checkpointing. The the variable hook for the output would do the right thing.

**Comment by t-vi** [comment]
Indeed, it's the same thing. Thank you for pointing this out!
